### PR TITLE
Rs/yj/rule6 7 (Previously 6-12)

### PR DESCRIPTION
**Pull request by yunjoonjung-PNNL**
I re-created 6-7 PR (previously 6-12) because I renamed the branch. I'd appreciate it if you reviewed this PR. Thank you!

### CODE REVIEW

File: rct229/rules/section6/section6rule7.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+class Section6Rule7(RuleDefinitionListIndexedBase):
+    """Rule 7 of ASHRAE 90.1-2019 Appendix G Section 6 (Lighting)"""
+
+    def __init__(self):
+        super(Section6Rule7, self).__init__(
+            rmrs_used=UserBaselineProposedVals(False, False, True),
+            each_rule=Section6Rule7.BuildingRule(),
+            index_rmr="proposed",
+            id="6-7",
+            description="User building is modeled with daylighting controls directly or through schedule adjustments.",
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rules/section6/section6rule7.py:22]: Fixed. Thanks for this finding.

File: rct229/rules/section6/section6rule7.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    class BuildingRule(RuleDefinitionListIndexedBase):
+        def __init__(self):
+            super(Section6Rule7.BuildingRule, self).__init__(
+                rmrs_used=UserBaselineProposedVals(False, False, True),
+                each_rule=Section6Rule7.BuildingRule.ZoneRule(),
+                index_rmr="proposed",
+                list_path="$..zones[*]",
+            )
+
+        class ZoneRule(RuleDefinitionBase):
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rules/section6/section6rule7.py:35]: It totally makes sense! Thanks for this kind comment.

**Comment by yunjoonjung-PNNL** [comment]
I reflected on all Weili's comments and would appreciate it if you could review this PR one more time. Thank you!

File: rct229/rules/section6/__init__.py [hunk 3]
```diff
@@ -7,10 +7,13 @@
     "section6rule3",
     "section6rule4",
     "section6rule6",
+    "section6rule7",
     "section6rule8",
     "section6rule9",
 ]
 
+from rct229.rules.section6 import section6rule7
```
Comment (charliepnnl) [inline rct229/rules/section6/__init__.py:15]: Was this being used during development and should not be removed?

File: rct229/rules/section6/section6rule7.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                    },
+                )
+
+            def get_calc_vals(self, context, data=None):
+                zone_p = context.proposed
+
+                daylight_flag_p = False
+                has_daylight_control_flag = False
+                for subsurface in find_all(
+                    "$..surfaces[?adjacent_to = 'EXTERIOR'].subsurfaces[?classification != 'DOOR']",
```
[the remark this answers is not in the document]
Comment (charliepnnl) [reply rct229/rules/section6/section6rule7.py]: I couldn't agree more!

File: rct229/rules/section6/section6rule7.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+                    },
+                    index_rmr="proposed",
+                    list_path="$..spaces[*]",
+                )
+
+            def create_data(self, context, data=None):
+                zone_p = context.proposed
+
+                return {
+                    "has_daylight_control_flag": len(
```
Comment (charliepnnl) [inline rct229/rules/section6/section6rule7.py:54]: According to the RDS, shouldn't this be `zone_pdaylight_flag_p`?
Otherwise, I love this part of the code.

File: rct229/rules/section6/section6rule7.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+                        find_all(
+                            "$..surfaces[?adjacent_to = 'EXTERIOR'].subsurfaces[?classification != 'DOOR']",
+                            zone_p,
+                        )
+                    )
+                    > 0
+                }
+
+            class SpaceRule(RuleDefinitionBase):
```
Comment (charliepnnl) [inline rct229/rules/section6/section6rule7.py:63]: It looks like the rule assertion cases in the RDS all refer to the zone. So I don't think we want a SpaceRule. Instead, I believe we want to:
- change the `create_data` above to `get_calc_vals`,
- move the calc_vals needed from `get_calc_vals` below to the `get_calc_vals` above at the zone level.

**Review by charliepnnl — CHANGES_REQUESTED** [review]
I believe BulidingRule and SpaceRule are not needed.

File: rct229/rules/section6/__init__.py (hunk 3 above)
Comment (yunjoonjung-PNNL) [reply rct229/rules/section6/__init__.py:15]: I don't know why this line was added here, but I really appreciate for pointing this out :).

File: rct229/rules/section6/section6rule7.py (hunk 5 above)
Comment (yunjoonjung-PNNL) [reply rct229/rules/section6/section6rule7.py:54]: Correct! I just realized that ```zone_pdaylight_flag_p``` should be ```daylight_flag_p``` in the RDS. Hence, I changed ```has_daylight_control_flag``` to ```daylight_flag_p```. I appreciate for finding this.

**Comment by yunjoonjung-PNNL** [comment]
I reflected on Charlie's comments and would appreciate it if you could review this PR one more time. Thank you!

File: rct229/rules/section6/section6rule7.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                        "$..surfaces[?adjacent_to = 'EXTERIOR'].subsurfaces[?classification != 'DOOR']",
+                        zone_p,
+                    )
+                )
+                > 0
+            )
+
+            has_daylight_control_flag = False
+            for lighting in find_all("$..interior_lighting[*]", zone_p):
+                interior_lighting_p = getattr_(
```
Comment (charliepnnl) [inline rct229/rules/section6/section6rule7.py:50]: I'm finding the intent of the RDS a little hard to follow with respect to `interior_lighting_p.are_schedules_used_for_modeling_daylighting_control` that appears in the Rule Assertion Cases 1 and 2.

The problem is that `interior_lighting_p.are_schedules_used_for_modeling_daylighting_control` refers to a specific space within the zone, but the Rule Assertions refer to the zone as a whole.
Is `interior_lighting_p.are_schedules_used_for_modeling_daylighting_control` intended as another flag which should be set if this is True for ANY space in the zone?

The implementation here resets `interior_lighting_p` in each iteration of the loop. So on the last one is returned.

**Review by charliepnnl — CHANGES_REQUESTED** [review]
I'm stuck on `interior_lighting_p.are_schedules_used_for_modeling_daylighting_control` - see my comment.

**Comment by yunjoonjung-PNNL** [comment]
I updated the code and RDS based on the feedback in Issue #567. I'd appreciate it if you could review it one more time. Thank you!

File: rct229/rules/section6/section6rule7.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                len(
+                    find_all(
+                        "$..surfaces[?adjacent_to = 'EXTERIOR'].subsurfaces[?classification != 'DOOR']",
+                        zone_p,
+                    )
+                )
+                > 0
+            )
+
+            has_daylight_control_flag = False
```
Comment (charliepnnl) [inline rct229/rules/section6/section6rule7.py:48]: Following the pattern for *daylight_flag_p* above, couldn't this be:
```
has_daylight_control_flag = (
        len(
            find_all(
                "$..spaces[*].interior_lighting[*][?daylighting_control_type != 'NONE']",
                zone_p,
            )
        )
        > 0
    )
```
and do something similar for *daylight_schedule_adjustment_flag*?

File: rct229/rules/section6/section6rule7.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+from rct229.rule_engine.rule_list_indexed_base import RuleDefinitionListIndexedBase
+from rct229.rule_engine.user_baseline_proposed_vals import UserBaselineProposedVals
+from rct229.utils.assertions import getattr_
+from rct229.utils.jsonpath_utils import find_all
+
+MSG_WARN_DAYLIGHT_NO_SCHEDULE = "Some of the spaces in zone are modeled with window(s) and/or skylight(s) and have daylighting controls modeled explicitly in the simulation tool. Verify that the mandatory lighting control requirements are met."
+MSG_WARN_DAYLIGHT = "Some of the spaces in zone are modeled with window(s) and/or skylight(s) and have daylighting controls modeled via schedule adjustment. Verify that the mandatory lighting control requirements are met, and that the supporting documentation is provided for the schedule adjustment."
+MSG_WARN_NO_DAYLIGHT = "Some of the spaces in zone are modeled with fenestration but no daylighting controls. The design must include mandatory daylighting controls unless any of the exceptions to 90.1 section 9.4.1.1€ apply."
+
+DAYLIGHT_CONTROL_TYPE = schema_enums["LightingDaylightingControlType"].NONE
```
Comment (charliepnnl) [inline rct229/rules/section6/section6rule7.py:12]: It might more self documenting to set 
```
DAYLIGHT_CONTROL_TYPE = schema_enums["LightingDaylightingControlType"]
```
here and then use DAYLIGHT_CONTROL_TYPE.NONE where this is used in the code.

File: rct229/rules/section6/section6rule7.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                    "are_schedules_used_for_modeling_daylighting_control",
+                )
+
+                if lighting["daylighting_control_type"] != DAYLIGHT_CONTROL_TYPE:
+                    has_daylight_control_flag = True
+
+            return {
+                "daylight_flag_p": daylight_flag_p,
+                "has_daylight_control_flag": has_daylight_control_flag,
+                "interior_lighting_p": interior_lighting_p,
```
Comment (charliepnnl) [inline rct229/rules/section6/section6rule7.py:62]: I don't think *interior_lighting_p* is used in the Rule Assertions any more, so I don't think this should be added to calc_vals.
But we do need *daylight_schedule_adjustment_flag*.

**Review by charliepnnl — COMMENTED** [review]
I think there is a mistake regarding *interior_lighting_p* vs *daylight_schedule_adjustment_flag* in *get_calc_vals*.

File: rct229/rules/section6/section6rule7.py (hunk 8 above)
Comment (yunjoonjung-PNNL) [reply rct229/rules/section6/section6rule7.py:48]: This is brilliant. I appreciate for letting me know about this :).

File: rct229/rules/section6/section6rule7.py (hunk 10 above)
Comment (yunjoonjung-PNNL) [reply rct229/rules/section6/section6rule7.py:62]: Fixed. Thanks for finding this and I should've been more meticulous.

File: rct229/rules/section6/section6rule7.py (hunk 9 above)
Comment (yunjoonjung-PNNL) [reply rct229/rules/section6/section6rule7.py:12]: Great point! This part was removed because we now use ```find_all``` function instead of for-loop.

**Review by charliepnnl — APPROVED** [review]
Looks good to me

File: rct229/rules/section6/section6rule7.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+                rmrs_used=UserBaselineProposedVals(False, False, True),
+            )
+
+        def get_calc_vals(self, context, data=None):
+            zone_p = context.proposed
+
+            daylight_flag_p = (
+                len(
+                    find_all(
+                        "$..surfaces[?adjacent_to = 'EXTERIOR'].subsurfaces[?classification != 'DOOR']",
```
Comment (weilixu) [inline rct229/rules/section6/section6rule7.py:39]: I will prefer to use enum instead of string here. See [5-24](https://github.com/pnnl/ruleset-checking-tool/blob/develop/rct229/rules/section5/section5rule24.py) as an example.
You can replace both `adjacent_to` and `classification' to the enum.

File: rct229/rules/section6/section6rule7.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+                        zone_p,
+                    )
+                )
+                > 0
+            )
+
+            has_daylight_control_flag = (
+                len(
+                    find_all(
+                        "$..spaces[*].interior_lighting[?daylighting_control_type!='NONE']",
```
Comment (weilixu) [inline rct229/rules/section6/section6rule7.py:49]: Use enum if possible.

File: rct229/rules/section6/section6rule7.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+            )
+
+            daylight_schedule_adjustment_flag = (
+                len(
+                    find_all(
+                        "$..spaces[*].interior_lighting[?are_schedules_used_for_modeling_daylighting_control == true]",
+                        zone_p,
+                    )
+                )
+                > 0
```
Comment (weilixu) [inline rct229/rules/section6/section6rule7.py:63]: Minor comment
You may also use any() function here (unless the `are_schedules_used_for_modeling_daylighting_control` is equal to a string, I don't quite remember how it is in the schema now) - this is a personal preference choice
```python
any(find_all("$..spaces[*].interior_lighting[*].are_schedules_used_for_modeling_daylighting_control", zone_p))
```

File: rct229/rules/section6/section6rule7.py [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+                "daylight_schedule_adjustment_flag"
+            ]
+
+            if daylight_flag_p and has_daylight_control_flag:
+                if daylight_schedule_adjustment_flag:
+                    manual_check_msg = MSG_WARN_DAYLIGHT
+                else:
+                    manual_check_msg = MSG_WARN_DAYLIGHT_NO_SCHEDULE
+
+            return manual_check_msg
```
Comment (weilixu) [inline rct229/rules/section6/section6rule7.py:91]: We tend to add the final logic to the return statement.

You do not need to worry about `if daylight_flag_p and has_daylight_control_flag:` because the `get_manual_check_required_msg` will only be called when `manual_check_required` is True.
This means, the `daylight_flag_p = True` and `has_daylight_control_flag = True` is the precondition for the `get_manual_check_required_msg` function in this case.

So you can rewrite this function to:
```python
def get_manual_check_required_msg(self, context, calc_vals=None, data=None):
    daylight_schedule_adjustment_flag = calc_vals[
                "daylight_schedule_adjustment_flag"
            ]
    return MSG_WARN_DAYLIGHT if  daylight_schedule_adjustment_flag else MSG_WARN_DAYLIGHT_NO_SCHEDULE
```

**Review by weilixu — APPROVED** [review]
Approved with minor comments for styling and coding preferences

File: rct229/rules/section6/section6rule7.py (hunk 14 above)
Comment (yunjoonjung-PNNL) [reply rct229/rules/section6/section6rule7.py:91]: Great point! I learned new thing :).

File: rct229/rules/section6/section6rule7.py (hunk 13 above)
Comment (yunjoonjung-PNNL) [reply rct229/rules/section6/section6rule7.py:63]: Great point. I reflected this in the code.

File: rct229/rules/section6/section6rule7.py (hunk 11 above)
Comment (yunjoonjung-PNNL) [reply rct229/rules/section6/section6rule7.py:39]: Great point! Fixed.

File: rct229/rules/section6/section6rule7.py (hunk 12 above)
Comment (yunjoonjung-PNNL) [reply rct229/rules/section6/section6rule7.py:49]: Fixed.

Comment (yunjoonjung-PNNL) [reply rct229/rules/section6/section6rule7.py:49]: Fixed.

File: rct229/rules/section6/section6rule7.py (hunk 11 above)
Comment (yunjoonjung-PNNL) [reply rct229/rules/section6/section6rule7.py:39]: Great point! Fixed.
